sdk/nodejs/runtime/resource.ts

```diff
@@ -91,18 +111,28 @@ export function readResource(res: Resource, t: string, name: string, props: Inpu
         // Now run the operation, serializing the invocation if necessary.
         const opLabel = `monitor.readResource(${label})`;
         runAsyncResourceOp(opLabel, async () => {
-            const resp: any = await debuggablePromise(new Promise((resolve, reject) =>
-                monitor.readResource(req, (err: Error, innerResponse: any) => {
-                    log.debug(`ReadResource RPC finished: ${label}; err: ${err}, resp: ${innerResponse}`);
-                    if (err) {
-                        preallocError.message =
-                            `failed to read resource #${resolvedID} '${name}' [${t}]: ${err.message}`;
-                        reject(preallocError);
-                    }
-                    else {
-                        resolve(innerResponse);
-                    }
-                })), opLabel);
+            let resp: any;
+            if (monitor) {
+                // If we're attached to the engine, make an RPC call and wait for it to resolve.
+                resp = await debuggablePromise(new Promise((resolve, reject) =>
+                    (monitor as any).readResource(req, (err: Error, innerResponse: any) => {
+                        log.debug(`ReadResource RPC finished: ${label}; err: ${err}, resp: ${innerResponse}`);
+                        if (err) {
+                            preallocError.message =
+                                `failed to read resource #${resolvedID} '${name}' [${t}]: ${err.message}`;
+                            reject(preallocError);
+                        }
+                        else {
+                            resolve(innerResponse);
+                        }
+                    })), opLabel);
+            } else {
+                // If we aren't attached to the engine, mock up a fake response for testing purposes.
+                resp = {
+                    getUrn: () => createUrn(t, name),
+                    getProperties: () => req.getProperties(),
+                };
```

Comment:
i think i'd like some way for this to be more intentional, rather than assuming we're in tests.  perhaps an internally exported 'runtime' entrypoint function that says "we're in test mode".

Reply:
Right now it warns if you don't set `PULUMI_DISCONNECTED=true`. What if we instead throw an error and only if you set `PULUMI_TEST_MODE=true`, is it allowed? Much more intentional.

Reply:
yes.  i would much prefer that :)

Reply:
agreed. please do that :)